### ISSUE TITLE
Allow condition meta

### ISSUE THREAD

**Issue by thejcannon** [fast-jsonapi/fast_jsonapi#66]
## Expected Behavior
The following to work
```ruby
meta if: Proc.new{...} do |record| {} end
```
(similar to conditional links)

## Actual Behavior
`ArgumentError (wrong number of arguments (given 1, expected 0)):`

**Comment by thejcannon** [comment]
Looking at the code, this isn't as easy as it was for links. Links use the `Scalar` class, which among other things, doesn't add to the hash given to `serialize` if the condition is present and fails.
For `meta`, the existence of the attribute means you'll end up with the key-value pair in the outer hash. You could do something similar to `Scalar`, but passing in the outer hash seems extreme. You could key off serialization being `nil`, but someone might want a `nil` meta value.

**Comment by stas** [comment]
@thejcannon adding support for the conditional would render a response without the key `meta` present if the condition is not met and defined. Is this what you are looking for?

Another important note here, is that we're talking about meta per resource and not per collection. Just to confirm the expectations.

**Comment by thejcannon** [comment]
@stas , yeah that's what I had in mind. FWIW the "condition" I'm using is if the instance is a collection, so I can avoid the `meta` key-value per-resource (as it might be expensive).

Right now, I'm just leaving the `meta` as `nil` and ignoring it.

**Comment by stas** [comment]
@thejcannon why not just use the meta option to solve your problem? Here's an example:
https://github.com/fast-jsonapi/fast_jsonapi#compound-document

You can use `jsonapi.rb` to get a nicer API to handle this, or pass it directly as part of your `render` call in `show/update/create`:
https://github.com/stas/jsonapi.rb#collection-meta

**Comment by thejcannon** [comment]
Personally, I like keeping the serialization logic in the serializer, and don't want it bleeding into the controller if I can help it. Allows me to look at the serializer and know what the output would look like without having to cross-check the controller.

**Comment by stas** [comment]
@thejcannon I think we're talking about two different things, please review the way serializer `MySerializer.new(meta: <stuff>)` work, you'd have to pass the meta either way before initializing the serializer class.

**Comment by thejcannon** [comment]
@stas Perhaps the confusion is that I'm using one serializer for both the collection and the instance.
 
When I'm using it for the instance I want a subset of the overall information, and I do that by adding conditions to the attributes and links that check if we're rendering a collection.
When I'm using it for the instance I want all the information.

I _could_ use the `meta` option when I'm rendering for the instance to _add_ it (and just leave it out of the serializer altogether) however that's moving serialization logic out of the serializer and into the controller.

Perhaps some (abbreviated) code would help
```ruby
class FooSerializer
  include FastJsonapi::ObjectSerializer
  attributes :baz, :bonk 

  attribute :bar, if: isnt_collection
  link :profile, if: isnt_collection do |x| '' end
  meta if: isnt_collection do |x|  {} end
```

**Comment by stas** [comment]
> @stas Perhaps the confusion is that I'm using one serializer for both the collection and the instance.

That's ok.

> When I'm using it for the collection I want a subset of the overall information, and I do that by adding conditions to the attributes and links that check if we're rendering a collection.
> When I'm using it for the instance I want all the information.

Did you consider using the [sparse fieldsets](https://github.com/fast-jsonapi/fast_jsonapi#sparse-fieldsets)?


> however that's moving serialization logic out of the serializer and into the controller

I'm not sure why you assume that. You can define a static method on the serializer class and call it from the controller. You still need to call the serializer from the controller, so I'm not sure why you see things so radical :)

**Comment by thejcannon** [comment]
> Did you consider using the sparse fieldsets?

I did at first, but couldn't get the secret sauce that would allow me to filter links as well.

> You still need to call the serializer from the controller, so I'm not sure why you see things so radical

I guess my view is that the conditional-ality of a field/link/meta should be consistent. If my fields/links are being filtered via an option in the serializer when I'm declaring them, I'd like meta to be as well. Conversely, if I'm using sparse fieldsets and conditionally passing meta as an option, I'd like to do the same for the links.

Past that, I think it's just personal preference to want to make the conditional-ality declaritive 😛 Easier on my handful of brain cells.

**Comment by stas** [comment]
@thejcannon definitely not trying to argue, so apologies for that. It just looks like you're doing something very custom instead of leveraging the JSONAPI spec (at least based on the alternative aproach against sparse fieldsets).

Back to being on-topic, any contribution is welcome :construction_worker_man:

**Comment by thejcannon** [comment]
@stas Well this is my very first `Ruby` project. So any discussion is beneficial 😄

If I find time, I'll definitely try my hand at a solution. At the present, I don't see an obvious, elegant solution.

**Comment by doits** [comment]
Just a quick try: Maybe you can accomplish a conditional meta already like this already?

```rb
meta do
  return if isnt_collection
  
  { ... }
end
```

**Comment by thejcannon** [comment]
@doits That's essentially what I'm doing today. The result isn't quite the same though.
Conditional meta would mean the output would either be `{..., "meta": ...}` or `{...}` (no meta at all).
Whereas `return if` means `{..., "meta": ...}` or `{..., "meta": null}` (always has the key)

**Comment by doits** [comment]
Ah OK, this looks a little bit messy and is not spec conform I assume? At least it skips the computation, but is no real solution then.

---

I have a proposal in #54 which makes `meta` work like `attribute`, which solves that problem and gives you much much more control about generating the meta hash. So it is much more than only adding the conditional.

Currently it does not look like this will be merged because the need for it is questioned, but I will show what you *could* do with it in your case.

```rb
class FooSerializer
  include FastJsonapi::ObjectSerializer

  # always output `meta: { key_always: ... }`
  meta :key_always { ... }

  # those keys will only be there *if* record is serialized as part of a collection,
  # so if `isnt_collection` is false it will be:
  # meta == { 'key_always': ..., 'key_for_collection': ..., 'another_key_for_collection': ... }
  with_options if: proc { !isnt_collection } do
    meta :key_for_collection { ... }
    meta :another_key_for_collection { ...}
  end

  # those keys will only be there *unless* record is part of a collection,
  # so if `isnt_collection` is true it will be:
  # meta == { 'key_always': ..., 'key_for_record': ..., 'another_key_for_record': ... }
  with_options if: proc { isnt_collection } do
    meta :key_for_record { ... }
    meta :another_key_for_record { ... }
  end
end
```

If this could help you, feel free to drop a vote in #54.

**Comment by stas** [comment]
> Currently it does not look like this will be merged because the need for it is questioned, but I will show what you _could_ do with it in your case.

@doits it's not fair what you're doing here! I understand that you're not happy about your proposal not being merged, but there's no need to share your unhappiness in every related conversation. :disappointed: 

I already suggested you to prepare a PR to solve this exact request and we'll get that part merged. There's no need to _advertise_ features nobody requested. Please be nice, we're all in the same boat here...

**Comment by doits** [comment]
@stas please explain what is not fair in pointing to a MR that solves this issue and explaining how it solves this issue, even if it does more than that.

If it is only the sentence you quoted that bothers you, I'm not exactly sure how I could have summarized the current state of #54 differently to give some context for @thejcannon - or do I see the current state of #54 wrong? Then please correct me.

But it is OK, I won't "advertise" this PR as a solution for issues anymore. Feel free to close #54, I'm OK with it and won't argue against it. I don't want to spend more time arguing for it.

**Comment by stas** [comment]
@doits it bothers me that a refusal to merge something brings the contributor to the idea that hijacking other conversations with one's frustration is ok. It's not ok, and yes, I think it's healthier for the sake of this project community to not receive contributions from such folks.

Thanks again for all your contributions though.

**Comment by thejcannon** [comment]
I went and looked at the spec: https://jsonapi.org/format/#document-meta

And I think what I want _isn't_ noncoforming. It says:
> Where specified, a meta member can be used to include non-standard meta-information.

The "where specified" tells me I'm allowed to omit it. That's good news.
But I also saw
> The value of each meta member MUST be an object

Which means explicitly or implicitly returning `nil` (`null`) won't cut it. So for now I'll just return an empty object.
